database.cc

```diff
@@ -1805,39 +1805,19 @@ future<mutation> database::apply_counter_update(schema_ptr s, const frozen_mutat
   }));
 }
 
-static future<> maybe_handle_reorder(std::exception_ptr exp) {
+future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db::timeout_clock::time_point timeout) {
+    db::rp_handle h;
+    if (cf.commitlog() != nullptr && cf.durable_writes()) {
+        auto fm = freeze(m);
+        commitlog_entry_writer cew(m.schema(), fm, db::commitlog::force_sync::no);
+        h = co_await cf.commitlog()->add_entry(m.schema()->id(), cew, timeout);
+    }
     try {
-        std::rethrow_exception(exp);
-        return make_exception_future(exp);
+        co_await apply_in_memory(m, cf, std::move(h), timeout);
     } catch (mutation_reordered_with_truncate_exception&) {
         // This mutation raced with a truncate, so we can just drop it.
         dblog.debug("replay_position reordering detected");
-        return make_ready_future<>();
-    }
-}
-
-future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db::timeout_clock::time_point timeout) {
-    if (cf.commitlog() != nullptr && cf.durable_writes()) {
-        return do_with(freeze(m), [this, &m, &cf, timeout] (frozen_mutation& fm) {
```

Comment:
A method with `do_with()` is a natural convert to coroutines without perf regression. Since the non-alloc case here is basically only used for tests AFAIK I don't think converting this method is a problem.